### PR TITLE
Database interface for multiple environments

### PR DESCRIPTION
**Pull request by logan-hendricks**
This change implements a database interface for multiple environments.

It does so by doing the following:

1. Establish a database interface for all interactions on the database
2. Allow for user overloading on different methods to enable multiple different drivers
3. Create a registry of potential database connections to select from
4. Establish an environment variable for users to select their database connection

For those unfamiliar with some of the changes implemented regarding structs & interfaces in go. These changes implement the GoLang strategy & singleton patterns

### CODE REVIEW

File: server/db/MockDB.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	players       map[string]model.Player
+}
+
+//type mongoDB struct {
+//	client   *mongo.Client
+//	uri      string
+//	database *mongo.Database
+//	games    *mongo.Collection
+//	players  *mongo.Collection
+//}
```
Comment (Aaron-Fine) [inline server/db/MockDB.go:23]: Can this commented out code be removed? If not, why not?

File: server/db/MockDB.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+// HasGameByID checks to see if a game with the given ID exists in the database.
+func (db *mockDB) HasGameByID(id string) bool {
+	_, ok := db.games[id]
+	return ok
+}
+
+// CreateGame a game with the given ID. Perhaps this should instead just return an id?
+func (db *mockDB) CreateGame() (*model.Game, error) {
+	myGame := model.Game{ID: uuid.New().String(), Password: "12234"}
```
Comment (Aaron-Fine) [inline server/db/MockDB.go:39]: Why is the password hardcoded?

File: server/db/UnoDB.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package db
+
+import (
+	"github.com/jak103/uno/model"
+)
+
+// UnoDB declares the database types for the applicaiton
+type UnoDB interface {
```
Comment (Aaron-Fine) [inline server/db/UnoDB.go:8]: As I understand it, this interface needs to have provisions for replacing all state updates from the `server/uno.go` file. This would mean we need ways to retrieve and modify a players hand for a given game, at the least. None of the existing interfaces appear to provide the needed functionality...

File: server/db/MockDB.go (hunk 2 above)
Comment (logan-hendricks) [reply server/db/MockDB.go:39]: This is a mockDB. it's specifically to be used for unit tests. The Game ID may be hardcoded at some point depending on how unit tests go long term.

File: server/db/MockDB.go (hunk 1 above)
Comment (logan-hendricks) [reply server/db/MockDB.go:23]: hmm... didn't notice that left in there. removing now.

File: server/db/UnoDB.go (hunk 3 above)
Comment (logan-hendricks) [reply server/db/UnoDB.go:8]: So.. to clarify my thoughts behind this. And this is where @hhenrichsen have had a difference of opinion.  I personally felt like the purpose of this task was to implement the `structure` in order to have an interface for different database connection types.  (ie, FireStore, Mongo, MockDB)

I personally didn't feel like the scope of this Issue was to actually `implement` all of the particular game functionality. There were other issues logged to ensure that can happen.

This particular interface can absolutely be expanded on by other team members. But If they don't know that the interface is there, or in place.... they won't add to it.  My concern is that we could spend all week rounding out all of the database reqeusts... and every other group working on a task will implement a database calls with some other paradigm in mind.  

TL;DR I agree that this interface doesn't implement all of the needed endpoints.... I just don't feel that was within the scope of this Task.. Which was in my opinion... to just implement the `stucture` of the database connection pattern.

Comment (jak103) [reply server/db/UnoDB.go:8]: In my mind, the DB interface should just be able to CRUD the game data. The uno.go code retrieves the game object, manipulates it and then stores it again. If we do as you are suggesting Aaron, then all the game functionality move into the database interface, which is not what we want.

Comment (jak103) [reply server/db/UnoDB.go:8]: I also see from the interface that it can also CRUD the player data. This is something I hadn't thought of, but is correct. We probably want to decouple the player and game data as we will want the player to exist beyond the scope of the game.
This will allow us to track stats and other info for the players. I like it.

**Comment by jak103** [comment]
Aaron, per my other comments, it is not the responsibility of the DB interface to know how to update every element of the game state. It is only responsible for store and retrieving the game and player states. Please remove your request for change.

**Comment by jak103** [comment]
I'll merge once @mattwhite180 , @Ryan-Johnson-1315 , and @hhenrichsen  approve their reviews.

**Comment by logan-hendricks** [comment]
> The registry looks good! I think I need to add Delete methods to the implementation before we merge, I can do that soon unless someone else wants to try.

I can potentially get to this after work. (but that's basically when class is starting). If you end up getting to it first that's fine. I'm personally not as concerned about getting `all` the functionality for the CRUD operations in place as I still think it can be expanded as features are built. However, I'm not opposed to placing those in.

**Comment by logan-hendricks** [comment]
@jak103 The one thing we did not implement on these changes is the actual implementation for the Firestore database. So all of this will work on Dev, but not on Prod. We were unsure how to move forward on that without your facilitation of credentials.

**Comment by jak103** [comment]
I didn't catch that there is no delete working. We probably should at least add it to the interface. I'm not concerned about the implementation right now as it isn't critical, but it should be on the interface.

**Comment by jak103** [comment]
I'll help work the Firestore in after I finish grading. I'll have to do some tags to get it out to test it, so lets get this PR merged first.

**Comment by hhenrichsen** [comment]
Go won't build without the structs implementing the interface, and it shouldn't be too bad to implement so I'm fine working on it when I add it.

**Comment by logan-hendricks** [comment]
@hhenrichsen I got a spare moment and work. and realized it was a pretty lightweight implementation on the mongo side of the delete as you already established the `Collections`. I just implemented stubs on the different structs, and a full implementation on the mongoDB driver. Let me know if you see any issues

File: server/db/FirebaseDB.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	return nil, errors.New("mockdb: game not found")
+}
+
+// LookupPlayer checks to see if a player is in the database
+func (db *firebaseDB) LookupPlayer(id string) (*model.Player, error) {
+	if player, ok := db.players[id]; ok {
+		return &player, nil
+	}
+	return nil, errors.New("mockdb: player not found")
+}
```
[the remark this answers is not in the document]
Comment (logan-hendricks) [reply server/db/FirebaseDB.go:81]: Ryan. Just to confirm your question. Are you asking about `LookupPlayer` being declared multiple times in the different files?

[the remark this answers is not in the document]
Comment (hhenrichsen) [reply server/db/FirebaseDB.go:81]: This just checking for a value in a map, and I understand the hesitation with it being repeated so much. We could write a function that looks for a key and returns `nil` if passed a map, but we still have the added complication of returning different types from each function and map. It really depends on where we want to spend our time.

**Comment by hhenrichsen** [comment]
> @hhenrichsen I got a spare moment at work. and realized it was a pretty lightweight implementation on the mongo side of the delete as you already established the `Collections`. I just implemented stubs on the different structs, and a full implementation on the mongoDB driver. Let me know if you see any issues

I went through and added the implementations, and also reordered the methods so they're together in the interface and classes. The mongo stuff looks good, thanks for doing that!

File: server/db/FirebaseDB.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (logan-hendricks) [reply server/db/FirebaseDB.go:81]: @Ryan-Johnson-1315 . To explain what I think your question is.  I apologize that this is going to wax a little verbose, but I just thought to error on the side of extra information than less information. If you happen to know all of this previously, I apologize.

What I imagine your question is, why are there so many many files have been created in the DB section. And why do some of these files appear to be mostly duplicates of each other.

So, in a lot of Object Oriented Languages, like Java, C++, C#, you have the option for method overloading on an abstract class.  In Go, there isn't the idea of a base class, or an abstract class. Basically we just define an interface, and then every struct that we establish, we need to entirely implement each method in order to honor the interface.

What this means is that you can't have a `default` way of handling things persay. you basically have to define each method individually (even if they are duplicates of each other)

However, the case of what we are trying to build now is somewhat uncommon.  We've established a single interface for all of our database interactions, and a series of subclasses that handle `different` types of database interactions. (Mongo, Firestore, and Mock/Unit tests).  

Typically  in most applications you maintain the same database type on dev as well as on prod. However, in our case we need to support `similar` database types, but not identical. Which means we need to translate application logic into database logic for each type of database. (So for example, normally you would just have one file to handle Database transactions, but if you are essentially supporting 3 different databases, you have to write 3 files, and establish a contract/interface that they are all required to honor) 

However... in our case, @hhenrichsen and I did not have all of the tools/items on hand to build out the specific database logic for interacting with Firestore. So we opted to currently have FirestoreDB be a duplicate of what our MockDB is accomplishing.  The MockDB is going to only be used for testing purposes, but it serves as a useful `stub` for our Firestore Connections until we finish those out.

Eventually after we get this shipped, we will work with @jak103 to actually round out the functionality for the FirestoreDB struct to be specific to Firestore itself.

**Comment by logan-hendricks** [comment]
> > @hhenrichsen I got a spare moment at work. and realized it was a pretty lightweight implementation on the mongo side of the delete as you already established the `Collections`. I just implemented stubs on the different structs, and a full implementation on the mongoDB driver. Let me know if you see any issues
> 
> I went through and added the implementations, and also reordered the methods so they're together in the interface and classes. The mongo stuff looks good, thanks for doing that!

Whatever, you were the hero on this particular PR. that was a herculean effort to get all of this implemented on your own in your free time. Well done :) Hopefully I didn't slow you down.

File: server/db/FirebaseDB.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ryan-Johnson-1315) [reply server/db/FirebaseDB.go:81]: Sounds good. I have been looking at a few implementations of how to do it it would behave as a overridden function, but I believe you are correct. I think this is the best way to do it. Everything else looks great!

**Comment by Aaron-Fine** [comment]
> Aaron, per my other comments, it is not the responsibility of the DB interface to know how to update every element of the game state. It is only responsible for store and retrieving the game and player states. Please remove your request for change.

I have removed my request for change. That said, I still don't see how to actually use this. From what I can tell, I'm expected to call `LookupGameByID` and get what I need from that. Yet all that seems to return is a `model.Game` which is only and ID and a password... Where is the list of players in a game? Where is the hand data for each player in that game? 

I'm hoping that I'm just not understanding what this interface is returning.

**Comment by logan-hendricks** [comment]
> > Aaron, per my other comments, it is not the responsibility of the DB interface to know how to update every element of the game state. It is only responsible for store and retrieving the game and player states. Please remove your request for change.
> 
> I have removed my request for change. That said, I still don't see how to actually use this. From what I can tell, I'm expected to call `LookupGameByID` and get what I need from that. Yet all that seems to return is a `model.Game` which is only an ID and a password... Where is the list of players in a game? Where is the hand data for each player in that game?
> 
> I'm hoping that I'm just not understanding what this interface is returning.

@Aaron-Fine good question. so the items in the `model` package are supposed to _eventually_ reflect what we want the database model to look like.  The advantage to mongoDB is that these data models can be updated as we flesh out features with very little impact on the application.

When we first started working on this issue, we started fleshing out **tons** of items on the models, and were taking a lot of time talking about database models, when the purpose of the Jira was just to create a structure for essentially different database drivers, ie.,... mongo, firebase, unit tests... etc

We had the understanding that these models will be fleshed out substantially in future features. But didn't feel it belonged in the scope of this task, and tried to just get the structure in a good place very quickly so as to support the other features being built.

**Comment by jak103** [comment]
Agreed. Right now there is very little in the game model. It wasn't the task of this issue to define the game model, but to create an interface to the game model. The game model needs to be fleshed out by the other groups.
